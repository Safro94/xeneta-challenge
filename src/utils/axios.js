import axios from 'axios';
import https from 'https';

const opts = {
	httpsAgent: new https.Agent({ rejectUnauthorized: false }),
	baseURL: process.env.REACT_APP_SERVER_URL,
};

const interceptorError = error => Promise.reject(error.response);

const instance = axios.create(opts);

instance.interceptors.response.use(response => response, interceptorError);
instance.defaults.headers.common['x-api-key'] = process.env.REACT_APP_API_KEY;

export default instance;
