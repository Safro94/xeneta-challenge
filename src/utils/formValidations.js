const validateSearchForm = ({
	departure,
	destination,
	departureDate,
	returnDate,
}) => {
	if (!departure || !destination || !departureDate) {
		return false;
	}

	if (departure.code === destination.code) {
		return false;
	}

	if (returnDate && returnDate < departureDate) {
		return false;
	}

	return true;
};

const errorCompare = (trip, message) => {
	const { departure, destination } = trip;

	const status = !departure && !destination ? true : false;

	return {
		message,
		status: !status && departure?.code === destination?.code,
	};
};

export { validateSearchForm, errorCompare };
