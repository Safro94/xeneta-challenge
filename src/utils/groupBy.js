import { marketPositions } from 'constants/marketPositions';

const groupBy = data =>
	data.reduce((accum, item) => {
		if (!accum[marketPositions.low]) accum[marketPositions.low] = [];
		if (!accum[marketPositions.average]) accum[marketPositions.average] = [];
		if (!accum[marketPositions.high]) accum[marketPositions.high] = [];

		accum[marketPositions.low].push({ x: item.day, y: item.low });
		accum[marketPositions.average].push({ x: item.day, y: item.mean });
		accum[marketPositions.high].push({ x: item.day, y: item.high });

		return accum;
	}, {});

export default groupBy;
