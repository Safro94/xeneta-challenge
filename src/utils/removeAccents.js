const removeAccents = value => {
	const map = {
		a: 'á|à|ã|â',
		e: 'é|è|ê',
		i: 'í|ì|î',
		o: 'ó|ò|ô|õ',
		u: 'ú|ù|û|ü',
	};

	for (let pattern in map) {
		value = value.replace(new RegExp(map[pattern], 'g'), pattern);
	}

	return value;
};

export { removeAccents };
