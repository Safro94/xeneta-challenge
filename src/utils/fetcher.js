import axios from 'utils/axios';

const fetcher = async ({ method = 'get', url, data = null }) => {
	return axios[method](url, JSON.parse(data)).then(
		response => response.data,
		error => {
			throw error;
		}
	);
};

export default fetcher;
