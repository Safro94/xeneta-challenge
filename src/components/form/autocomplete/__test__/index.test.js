import { render, screen } from 'utils/test-utils';

import userEvent from '@testing-library/user-event';

import Autocomplete from '../';

describe('Autocomplete', () => {
	const characters = 3;
	const source = [{ text: 'hello' }, { text: 'hello world' }];
	const error = {
		status: false,
		message: 'error',
	};
	const onSelectItem = jest.fn();
	const id = 'test';

	describe('Autocomplete snapshot', () => {
		it('should render OK', () => {
			const { container } = render(
				<Autocomplete error={error} source={source} characters={characters} />
			);
			expect(container).toMatchSnapshot();
		});
	});

	describe('Autocomplete functionality', () => {
		it('should show suggestions after the user enters 3 characters', () => {
			render(
				<Autocomplete
					onSelectItem={onSelectItem}
					error={error}
					characters={characters}
					source={source}
					id={id}
				/>
			);

			const input = screen.getByRole('textbox');
			userEvent.type(input, 'Hello');

			expect(screen.getAllByRole('listbox')).toHaveLength(2);
			expect(screen.getAllByText(/hello/i)).toHaveLength(2);
			expect(screen.getByText(/world/i)).toBeInTheDocument();

			userEvent.type(input, 'Hello world');
			expect(screen.getAllByText(/hello/i)).toHaveLength(1);

			userEvent.click(screen.getAllByTestId('menu-item')[0]);
			expect(onSelectItem).toHaveBeenCalledTimes(1);
			expect(onSelectItem).toHaveBeenCalledWith({ text: 'hello world' }, id);
		});

		it('should show the error', () => {
			const error = {
				status: true,
				message: 'error',
			};

			render(
				<Autocomplete
					onSelectItem={onSelectItem}
					error={error}
					characters={characters}
					source={source}
					id={id}
				/>
			);

			expect(screen.getByText(error.message)).toBeInTheDocument();
		});
	});
});
