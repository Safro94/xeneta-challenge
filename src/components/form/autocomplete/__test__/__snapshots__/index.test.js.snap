// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Autocomplete snapshot should render OK 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="react-autowhatever-1"
    class="container"
    role="combobox"
  >
    <div>
      <input
        aria-autocomplete="list"
        aria-controls="react-autowhatever-1"
        autocomplete="off"
        class="input"
        data-testid="input-text"
        type="text"
        value=""
      />
    </div>
    <div
      id="react-autowhatever-1"
      role="listbox"
    />
  </div>
</div>
`;
